### PR TITLE
Move templates to generic web host

### PR DESCRIPTION
**Pull request by Tratcher**
#4187 I started with one project, but it should be the same for all of them.

We'll also need to update the EF tooling for the new pattern.

### CODE REVIEW

File: src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs [hunk 1]
```diff
@@ -14,11 +15,14 @@ public class Program
...
         {
-            CreateWebHostBuilder(args).Build().Run();
+            CreateHostBuilder(args).Build().Run();
         }
 
-        public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
-            WebHost.CreateDefaultBuilder(args)
-                .UseStartup<Startup>();
+        public static IHostBuilder CreateHostBuilder(string[] args) =>
+            Host.CreateDefaultBuilder(args)
```
Comment (DamianEdwards) [inline src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs]: Still not 100% sold on `Host`. Perhaps `AppHost`?

File: src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs [hunk 2]
```diff
@@ -14,11 +15,14 @@ public class Program
...
         public static void Main(string[] args)
         {
-            CreateWebHostBuilder(args).Build().Run();
+            CreateHostBuilder(args).Build().Run();
         }
 
-        public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
-            WebHost.CreateDefaultBuilder(args)
-                .UseStartup<Startup>();
+        public static IHostBuilder CreateHostBuilder(string[] args) =>
```
Comment (DamianEdwards) [inline src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs]: Not for this PR, but I'd like us to revisit the `CreateHostBuilder` pattern completely in 3.0, i.e. question whether we still think it's worth the potential confusion of an untyped, pattern-based alternate entry point.

File: src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs (hunk 1 above)
Comment (davidfowl) [reply src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs]: Yea Host is hella generic but this is the generic host 😬. We should do an experience review.

File: src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs (hunk 2 above)
Comment (davidfowl) [reply src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs]: Agreed. I think EF still relies on this though.

**Comment by Tratcher** [comment]
@Eilon Does someone handle the F# for us? I could take a stab at it if needs be.

File: src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs (hunk 1 above)
Comment (davidfowl) [reply src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs]: AppHost I don’t like but that isn’t a hill I would die on.

- Host
- AppHost
- ApplicationHost
- GenericHost

**Comment by Eilon** [comment]
We do the F# ourselves, but we can have the F# experts review it if you're not certain it's both correct & idiomatic.

**Comment by Tratcher** [comment]
Ready for review. Pay attention to the F# code. I did what I could to verify it, but it's not currently possible to build and test the templates locally.

The test failures are known unrelated issues.

File: src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-FSharp/Program.fs [hunk 3]
```diff
@@ -8,18 +8,20 @@ open System.Threading.Tasks
 open Microsoft.AspNetCore
 open Microsoft.AspNetCore.Hosting
 open Microsoft.Extensions.Configuration
+open Microsoft.Extensions.Hosting;
```
Comment (muratg) [inline src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-FSharp/Program.fs]: Remove semicolon

File: src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-FSharp/Program.fs [hunk 4]
```diff
@@ -8,18 +8,20 @@ open System.Threading.Tasks
...
 
-    let CreateWebHostBuilder args =
-        WebHost
-            .CreateDefaultBuilder(args)
-            .UseStartup<Startup>();            
+    let CreateHostBuilder args =
+        Host.CreateDefaultBuilder(args)
+            .ConfigureWebHostDefaults(fun webBuilder ->
+                webBuilder.UseStartup<Startup>() |> ignore
+            );
```
Comment (muratg) [inline src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-FSharp/Program.fs]: Semicolon probably not needed

File: src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/StarterWeb-FSharp/Program.fs [hunk 5]
```diff
@@ -8,18 +8,20 @@ open System.Threading.Tasks
 open Microsoft.AspNetCore
 open Microsoft.AspNetCore.Hosting
 open Microsoft.Extensions.Configuration
+open Microsoft.Extensions.Hosting;
```
Comment (muratg) [inline src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/StarterWeb-FSharp/Program.fs]: Ditto for this file

File: src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/WebApi-FSharp/Program.fs [hunk 6]
```diff
@@ -8,18 +8,20 @@ open System.Threading.Tasks
 open Microsoft.AspNetCore
 open Microsoft.AspNetCore.Hosting
 open Microsoft.Extensions.Configuration
+open Microsoft.Extensions.Hosting;
```
Comment (muratg) [inline src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/WebApi-FSharp/Program.fs]: And this.

**Review by muratg — COMMENTED** [review]
Only reviewed F# changes, they look good except for the comments below.

It would be good to build the templates though. Why is it not possible now? What's blocking?

**Comment by Tratcher** [comment]
Right now the build is very unstable with a mixture of different build designs. Building Templating requires building the entire repo with several additional flags. https://github.com/aspnet/AspNetCore/pull/5958 should help.

**Comment by Eilon** [comment]
@cartermp - could you have a quick look at these ASP.NET Core template changes for the F# template?

**Review by cartermp — APPROVED** [review]
F# changes LGTM - I'd pull them down to build/run myself but it sounds like local template builds are challenging :)

**Comment by Eilon** [comment]
Thanks @cartermp !

**Review by davidfowl — APPROVED** [review]
@Tratcher is this ready to merge now?

**Comment by Tratcher** [comment]
Yes. I keep rebasing to get a cleaner test pass but...

**Comment by davidfowl** [comment]
I do think this is problematic https://github.com/aspnet/AspNetCore/issues/5960 and we can't update the templates without a replacement.

**Comment by Tratcher** [comment]
Yes #5960 needs to be fixed within the same preview release, but not before this checkin.

**Comment by Tratcher** [comment]
Actually, WebApplicationFactory can wait, it's only WebHostFactoryResolver that really needs to be done in the same preview or else the EF tooling won't work.

**Comment by davidfowl** [comment]
No, both things need to work. That's currently the recommended way to unit test MVC applications. What does the upgrade story look like if you have a bunch of tests using the WebHostBuilder today? Do we tell those users to keep using the WebHostBuilder?

**Comment by bitbonk** [comment]
Is there a preview build of the SDK that lets my try this yet?

**Comment by davidfowl** [comment]
Not yet.

**Pull request merged**